A couple PEP 654 edits (#1830)
python/peps: pep-0654.rst

==========

The interpreter is currently able to propagate at most one exception at a
time. The chaining features introduced in :pep:`3134` link together
exceptions that are related to each other as the cause or context, but
there are situations where multiple unrelated exceptions need to be propagated
together as the stack unwinds. Several real world use cases are listed below.

* **Concurrent errors**. Libraries for async concurrency provide APIs to invoke
  multiple tasks and return their results in aggregate. There isn't currently
  a good way for such libraries to handle situations where multiple tasks
  raise exceptions. The Python standard library's ``asyncio.gather()`` [1]_
  function provides two options: raise the first exception, or return the
  exceptions in the results list.  The Trio [2]_
  library has a ``MultiError`` exception type which it raises to report a

>   collection of errors. Work on this PEP was initially motivated by the

Reviewer: Maybe also link to `[9]` here?

  difficulties in handling ``MultiErrors``, which are detailed in a design
  document for an improved version, ``MultiError2`` [3]_.
  That document demonstrates how difficult it is to create an effective API
  for reporting and handling multiple errors without the language changes we
  are proposing (see also the `Programming Without except*`_ section.)

* **Implementing better Task spawning API in asyncio**.
  The current asyncio API for spawning child Tasks -- the ``asyncio.gather()``
  function [1]_ -- has a few shortcomings, the most notable of which is
  how it handles exceptions.  A better API, inspired by the Trio's "Nursery"
  primitive [13]_, has been a long-awaited feature of asyncio, blocked on

>   Python not having native language level support for exception groups.

Reviewer: This needs some more elaboration. Maybe just reference the complexity of Trio's solution?

Author: @iritkatriel @gvanrossum I can add a section to the PEP about TaskGroups API. Or you guys think that creating a bpo with the proposed design would be a better move?

Reviewer 2: Adding that design here means we are asking the steering council to approve it as part of this PEP, and if it's not straightforward I think it makes this PEP too big in scope. To Guido's point - isn't this really the same problem/use case as Trio's?

Reviewer: Yeah, I think we could shorten this parargaph so it just says "the same applies for asyncio".

Author: My problem with removing every notion of TaskGroups from this PEP weakens its motivation.  Ultimately adding task groups to asyncio were the main reason why we started working on this PEP.

Reviewer: So say exactly that.

Reviewer 2: I don't think you should remove it, just roll it into the same paragraph with Trio because it's the same kind of use case.  Like the paragraph about multiple use callbacks has two different use cases (atexit and pytest).

Author: I've pushed an update, basically adding this paragraph to the first bullet point:

>   Implementing a better, inspired by Trio nurseries, task spawning API in
  asyncio was the main motivation for this PEP. That work is currently blocked
  on Python not having native language level support for exception groups.

I propose to merge this PR with that, but if you guys want to phrase it differently please open a follow up PR.

>   Python not having native language level support for exception groups.

Reviewer 2: Is there an open issue or something else you could link to here?

Grouping several exceptions together can be done without changes to the
language, simply by creating a container exception type.
Trio [2]_ is an example of a library that has made use of this technique in its

> ``MultiError`` [9]_ type. However, such approaches require calling code to catch
> the container exception type, and then inspect it to determine the types of
> errors that had occurred, extract the ones it wants to handle and reraise the
> rest.

Reviewer: ```suggestion
``MultiError`` [9]_ type. However, such an approache requires calling code to catch
the container exception type, and then to inspect it to determine the types of
errors that had occurred, extract the ones it wants to handle, and reraise the
rest.
```

are handling ``ExceptionGroups`` of ``T`` rather than only naked ``Ts`` should be
specified in the same place where we state ``T``.

> Programming Without except*

Reviewer: ```suggestion
Programming Without ``except *``
```
(Not sure if RsT likes code blocks in titles though.)

Author: It should handle them OK, but they will look awkward when rendered to HTML. So I personally try to avoid using them in titles.

Reviewer: Maybe use single quotes then?

Author: Alright

===========================

> Consider the following simple example of the ``except *`` syntax (suppose
> Trio natively supports this proposal):

Reviewer: ```suggestion
Consider the following simple example of the ``except *`` syntax (pretending
Trio natively supported this proposal):
```

(Also: it's a little unconventional to see the "efficient" version first and the "inefficient" version second. Usually we do before/after, where before is the old way and after is the new, improved way. :-)

Author: Yeah! I had the same thought, but then opted for my variant because of just how arcane the "inefficient" way looks like compared to `except*`. I'm afraid that readers would be lost in it and skip the whole section, not getting why we added it in the first place. :/

.. code-block::

   try:
     async with trio.open_nursery() as nursery:
       # Make two concurrent calls to child()
       nursery.start_soon(child)
       nursery.start_soon(child)
   except *ValueError:

>      pass

Reviewer: Please use 4-space indents.

Author: Done

Here is how this code would look in Python 3.9:

.. code-block::

   def handle_ValueError(exc):
     if isinstance(exc, ValueError):
       return None
     else:

>        return exc

Reviewer 2: ```suggestion
       return exc   # reraise exc
```

   with MultiError.catch(handle_ValueError):
     async with trio.open_nursery() as nursery:
       # Make two concurrent calls to child()
       nursery.start_soon(child)
       nursery.start_soon(child)

This example clearly demonstrates how unintuitive and cumbersome handling
of multiple errors is in current Python.  The exception handling logic has
to be in a separate closure and is fairly low level, requiring the writer to
have non-trivial understanding of both Python exceptions mechanics and the
Trio APIs.  Instead of using the ``try..except`` block we have to use a

> ``with`` block.  Handling more exception types or implementing more complex

Reviewer 2: ```suggestion
``with`` block.  We need to explicitly reraise exceptions we are not handling. Handling more exception types or implementing more complex
```

> ``with`` block.  We need to explicitly reraise exceptions we are not handling. Handling more exception types or implementing more complex
> exception handling logic will only further complicate the code to the point

Reviewer: ```suggestion
``with`` block.  We need to explicitly reraise exceptions we are not handling.
Handling more exception types or implementing more complex
exception handling logic will only further complicate the code to the point
```
